src/Middleware/RequestThrottling/test/MiddlewareTests.cs

```diff
@@ -127,5 +127,38 @@ public void MultipleRequestsFillUpQueue()
             _ = middleware.Invoke(new DefaultHttpContext());
             Assert.Equal(2, middleware.ActiveRequestCount);
         }
+
+        [Fact]
+        public async void FullQueueInvokesOnRejected()
+        {
+            bool onRejectedInvoked = false;
+
+            var middleware = TestUtils.CreateTestMiddleware(
+                maxConcurrentRequests: 0,
+                requestQueueLimit: 0,
+                onRejected: httpContext =>
+                {
+                    onRejectedInvoked = true;
+                    return Task.CompletedTask;
+                });
+
+            var context = new DefaultHttpContext();
+            await middleware.Invoke(context).OrTimeout();
+            Assert.True(onRejectedInvoked);
+        }
+
+        [Fact]
+        public async void ExceptionThrownDuringOnRejected()
+        {
+            var middleware = TestUtils.CreateTestMiddleware(
+                maxConcurrentRequests: 0,
```

Comment:
This is a good start, but can we expand this test a little? I recommend the following.

1. Make `maxConcurrentRequests` equal to 1
2. Pass in a `next` RequestDelegate that returns the Task from a TCS (TaskCompletionSource) that you define at the top of the test.
3. Invoke the middleware and store the returned Task without awaiting it.
4. Invoke the middleware a second time and assert the DivideByZeroException like you're already doing.
5. Complete the TCS used by the `next`.
6. Assert that the Task returned from the first middleware invocation completes successfully.
7. Invoke the middleware a third time and Assert that it completes successfully since there's no other active requests and the Task returned by `next` is now completed.

You can also assert the `ActiveRequestCount` values like @DylanDmitri suggested in a previous comment.

Reply:
Another behavior change: we are afraid that people will overwrite `OnRejected` to log a message, and then the server will return an http 200.

The fix here is having the middleware set the statuscode to 503, and then invoking the user provided callback. The user can still set custom status codes, but by default it 503s.

Also, this means that the default `OnRejected` (in the options file) is now an empty lambda, since its current functionality gets moved to the middleware.